src/file-manager/fm-desktop-wayland-bg-dialog.c

```diff
@@ -0,0 +1,454 @@
+    static GtkCssProvider *provider;
+    gchar *css;
+    GString *string;
+    static GdkRectangle geometry = {0};
+
+    /* setup the preview only once*/
+    if (starting == TRUE)
+    {
+        static GtkWidget *preview;
+
+        /*Get the size and shape of the desktop*/
+        GdkDisplay *display = gdk_screen_get_display (gdk_screen_get_default());
+        GdkMonitor *monitor = gdk_display_get_monitor (display, 0);
+        gdk_monitor_get_geometry (monitor, &geometry);
+
+        preview = gtk_box_new (GTK_ORIENTATION_VERTICAL, 0);
+        gtk_widget_set_size_request (preview, geometry.width / 5, geometry.height / 5);
+        gtk_widget_set_name (GTK_WIDGET (preview), "caja-wayland-bg-preview");
+
+        preview_image = gtk_image_new ();
+        provider = gtk_css_provider_new ();
+        gtk_style_context_add_provider (gtk_widget_get_style_context (preview),
+                                        GTK_STYLE_PROVIDER (provider),
+                                        GTK_STYLE_PROVIDER_PRIORITY_APPLICATION);
+        gtk_box_pack_start (GTK_BOX (preview), preview_image, FALSE, FALSE, 0);
+        gtk_box_pack_start (GTK_BOX (box), preview, FALSE, FALSE, 0);
+    }
+
+    /* No image filename means we are previewing a color or gradient background*/
+    if ((!filename) || (strcmp (filename, "") == 0) || (strcmp (filename, " ") == 0))
+
+    {
+        gtk_image_clear (GTK_IMAGE(preview_image));
+
+        /*Build a color preview using a cssprovider due to requirement to handle RBGA values*/
+        string = g_string_new(NULL);
+        g_string_append (string, "#caja-wayland-bg-preview {");
+
+        if (strcmp (shading_type, "vertical-gradient") == 0)
+            g_string_append (string, "background-image: linear-gradient(to bottom,");
+
+        else if  (strcmp (shading_type, "horizontal-gradient") == 0)
+            g_string_append (string, "background-image: linear-gradient(to right,");
+
+        else
+        {
+            g_string_append (string, "background-color:");
+            g_string_append (string, primary_color_str);
+        }
+
+        if ((strcmp (shading_type, "vertical-gradient") == 0) ||
+           (strcmp (shading_type, "horizontal-gradient") == 0))
+        {
+            g_string_append (string, primary_color_str);
+            g_string_append (string, ",");
+            g_string_append (string, secondary_color_str);
+            g_string_append (string, ");");
+        }
+            g_string_append (string, "}");
+
+        css = g_string_free (string, FALSE);
+
+        gtk_css_provider_load_from_data (provider, css, -1, NULL);
+
+        g_free (css);
+
+    }
+    else
+    /*Preview a background image*/
+    {
+        GdkPixbuf  *pixbuf;
+
+        pixbuf = gdk_pixbuf_new_from_file_at_scale (filename, geometry.width / 5,
+                                                    geometry.height / 5, TRUE, NULL);
+
+        gtk_image_set_from_pixbuf (GTK_IMAGE (preview_image), pixbuf);
```

Comment:
This is very dangerous because the ```preview_image``` may not have been initialized

Reply:
Again, we have as line 80
        preview_image = gtk_image_new ();
in the     if (starting == TRUE)
block of `update_preview`
For this case I can add a check so that if the image is not valid we create it, or if that would lead to multiple copies simply return/.